File: tests/test_rest_api/test_runs_endpoint.py
"""
Tests for the GET /api/v1/runs endpoints.

Tests cover:
- GET /api/v1/runs (list all runs)
- GET /api/v1/runs/{run_id} (get specific run)
- GET /api/v1/runs/{run_id}/latest (get latest step)
"""

import pytest
import asyncio
from fastapi.testclient import TestClient

from tests.fixtures.metric_fixtures import valid_minimal_payload, valid_three_layer_network


class TestListAllRuns:
    """Tests for GET /api/v1/runs endpoint."""

    def test_empty_store_returns_empty_dict(self, client: TestClient):
        """Test that empty store returns empty dictionary."""
        response = client.get("/api/v1/runs")
        assert response.status_code == 200
        assert response.json() == {}

    def test_list_runs_after_single_submission(self, client: TestClient):
        """Test listing runs after submitting one metric."""
        client.post("/api/v1/metrics/layerwise", json=valid_minimal_payload())
        response = client.get("/api/v1/runs")
        assert response.status_code == 200
        data = response.json()
        assert "minimal_test" in data
        assert isinstance(data["minimal_test"], dict)

    def test_list_runs_response_structure(self, client: TestClient):
        """Test that list runs response has correct structure."""
        client.post("/api/v1/metrics/layerwise", json=valid_minimal_payload())
        response = client.get("/api/v1/runs")
        data = response.json()
        run_info = data["minimal_test"]
        assert "created_at" in run_info
        assert "last_update" in run_info
        assert "step_count" in run_info
        assert "latest_step" in run_info

    def test_step_count_is_one_after_single_submission(self, client: TestClient):
        """Test that step_count is 1 after single submission."""
        client.post("/api/v1/metrics/layerwise", json=valid_minimal_payload())
        response = client.get("/api/v1/runs")
        data = response.json()
        assert data["minimal_test"]["step_count"] == 1

    def test_latest_step_matches_submission(self, client: TestClient):
        """Test that latest_step matches the submitted step."""
        payload = valid_minimal_payload()
        step = 250
        payload["metadata"]["global_step"] = step
        client.post("/api/v1/metrics/layerwise", json=payload)
        response = client.get("/api/v1/runs")
        data = response.json()
        assert data["minimal_test"]["latest_step"] == step

    def test_multiple_runs_all_listed(self, client: TestClient):
        """Test that multiple runs are all listed."""
        run_ids = ["run_1", "run_2", "run_3"]
        for run_id in run_ids:
            payload = valid_minimal_payload()
            payload["metadata"]["run_id"] = run_id
            client.post("/api/v1/metrics/layerwise", json=payload)
        response = client.get("/api/v1/runs")
        data = response.json()
        for run_id in run_ids:
            assert run_id in data

    def test_created_at_format(self, client: TestClient):
        """Test that created_at is in ISO format."""
        client.post("/api/v1/metrics/layerwise", json=valid_minimal_payload())
        response = client.get("/api/v1/runs")
        data = response.json()
        created_at = data["minimal_test"]["created_at"]
        assert isinstance(created_at, str)
        assert "T" in created_at  # ISO format includes T

    def test_latest_step_null_for_new_run(self, client: TestClient, mock_store):
        """Test that latest_step is null for a run with no steps yet."""
        # Create a run without adding steps
        asyncio.run(mock_store.add_metrics(
            __import__('main').MetricsPayload(**valid_minimal_payload())
        ))
        response = client.get("/api/v1/runs")
        data = response.json()
        # Note: After adding metrics, there will be one step, so latest_step won't be null
        # This test verifies the structure exists
        assert "latest_step" in data["minimal_test"]


class TestGetSpecificRun:
    """Tests for GET /api/v1/runs/{run_id} endpoint."""

    def test_get_existing_run_returns_200(self, client: TestClient):
        """Test getting an existing run returns 200."""
        client.post("/api/v1/metrics/layerwise", json=valid_minimal_payload())
        response = client.get("/api/v1/runs/minimal_test")
        assert response.status_code == 200

    def test_get_nonexistent_run_returns_404(self, client: TestClient):
        """Test getting a nonexistent run returns 404."""
        response = client.get("/api/v1/runs/nonexistent_run")
        assert response.status_code == 404

    def test_get_nonexistent_run_error_format(self, client: TestClient):
        """Test that 404 error has correct format."""
        response = client.get("/api/v1/runs/nonexistent_run")
        data = response.json()
        assert "detail" in data
        assert isinstance(data["detail"], dict)
        assert data["detail"]["error"] == "not_found"

    def test_get_nonexistent_run_error_message(self, client: TestClient):
        """Test that 404 error includes run name in message."""
        response = client.get("/api/v1/runs/nonexistent_run")
        data = response.json()
        assert "nonexistent_run" in data["detail"]["message"]

    def test_get_run_response_structure(self, client: TestClient):
        """Test that get run response has correct structure."""
        client.post("/api/v1/metrics/layerwise", json=valid_minimal_payload())
        response = client.get("/api/v1/runs/minimal_test")
        data = response.json()
        assert "created_at" in data
        assert "last_update" in data
        assert "steps" in data
        assert isinstance(data["steps"], list)

    def test_get_run_steps_structure(self, client: TestClient):
        """Test that steps in run have correct structure."""
        client.post("/api/v1/metrics/layerwise", json=valid_minimal_payload())
        response = client.get("/api/v1/runs/minimal_test")
        data = response.json()
        assert len(data["steps"]) > 0
        step = data["steps"][0]
        assert "step" in step
        assert "timestamp" in step
        assert "batch_size" in step
        assert "layers" in step
        assert "cross_layer" in step

    def test_get_run_multiple_steps(self, client: TestClient):
        """Test getting a run with multiple steps."""
        run_id = "multi_step_test"
        for step_num in [100, 200, 300]:
            payload = valid_minimal_payload()
            payload["metadata"]["run_id"] = run_id
            payload["metadata"]["global_step"] = step_num
            client.post("/api/v1/metrics/layerwise", json=payload)
        response = client.get("/api/v1/runs/multi_step_test")
        data = response.json()
        assert len(data["steps"]) == 3

    def test_get_run_steps_sorted(self, client: TestClient):
        """Test that steps are sorted by step number."""
        run_id = "sorted_test"
        # Submit in reverse order
        for step_num in [300, 100, 200]:
            payload = valid_minimal_payload()
            payload["metadata"]["run_id"] = run_id
            payload["metadata"]["global_step"] = step_num
            client.post("/api/v1/metrics/layerwise", json=payload)
        response = client.get("/api/v1/runs/sorted_test")
        data = response.json()
        step_numbers = [step["step"] for step in data["steps"]]
        assert step_numbers == [100, 200, 300]

    def test_get_run_layers_structure(self, client: TestClient):
        """Test that layer data has correct structure."""
        client.post("/api/v1/metrics/layerwise", json=valid_three_layer_network())
        response = client.get("/api/v1/runs/three_layer_net")
        data = response.json()
        step = data["steps"][0]
        assert len(step["layers"]) == 3
        layer = step["layers"][0]
        assert "layer_id" in layer
        assert "layer_type" in layer
        assert "depth_index" in layer
        assert "intermediate_features" in layer
        assert "gradient_flow" in layer
        assert "parameter_statistics" in layer


class TestGetLatestStep:
    """Tests for GET /api/v1/runs/{run_id}/latest endpoint."""

    def test_get_latest_step_returns_200(self, client: TestClient):
        """Test getting latest step returns 200."""
        client.post("/api/v1/metrics/layerwise", json=valid_minimal_payload())
        response = client.get("/api/v1/runs/minimal_test/latest")
        assert response.status_code == 200

    def test_get_latest_step_nonexistent_run_returns_404(self, client: TestClient):
        """Test getting latest step from nonexistent run returns 404."""
        response = client.get("/api/v1/runs/nonexistent_run/latest")
        assert response.status_code == 404

    def test_get_latest_step_response_structure(self, client: TestClient):
        """Test that latest step response has correct structure."""
        client.post("/api/v1/metrics/layerwise", json=valid_minimal_payload())
        response = client.get("/api/v1/runs/minimal_test/latest")
        data = response.json()
        assert "step" in data
        assert "timestamp" in data
        assert "batch_size" in data
        assert "layers" in data
        assert "cross_layer" in data

    def test_get_latest_step_returns_actual_latest(self, client: TestClient):
        """Test that latest endpoint returns the most recent step."""
        run_id = "latest_test"
        for step_num in [100, 200, 300]:
            payload = valid_minimal_payload()
            payload["metadata"]["run_id"] = run_id
            payload["metadata"]["global_step"] = step_num
            client.post("/api/v1/metrics/layerwise", json=payload)
        response = client.get("/api/v1/runs/latest_test/latest")
        data = response.json()
        assert data["step"] == 300

    def test_get_latest_step_with_single_step(self, client: TestClient):
        """Test getting latest step when there's only one step."""
        client.post("/api/v1/metrics/layerwise", json=valid_minimal_payload())
        response = client.get("/api/v1/runs/minimal_test/latest")
        data = response.json()
        assert data["step"] == 100

    def test_get_latest_step_after_duplicate_submission(self, client: TestClient):
        """Test that duplicate step submission updates the latest step."""
        run_id = "duplicate_latest_test"
        payload = valid_minimal_payload()
        payload["metadata"]["run_id"] = run_id
        payload["metadata"]["global_step"] = 100

        # First submission
        client.post("/api/v1/metrics/layerwise", json=payload)

        # Modify and submit same step again
        payload["layer_statistics"][0]["intermediate_features"]["activation_std"] = 0.5
        client.post("/api/v1/metrics/layerwise", json=payload)

        # Get latest and verify it was updated
        response = client.get("/api/v1/runs/duplicate_latest_test/latest")
        data = response.json()
        assert data["step"] == 100
        # The updated value should be present
        assert data["layers"][0]["intermediate_features"]["activation_std"] == 0.5


class TestHealthEndpoint:
    """Tests for GET /health endpoint."""

    def test_health_returns_200(self, client: TestClient):
        """Test that health endpoint returns 200."""
        response = client.get("/health")
        assert response.status_code == 200

    def test_health_response_structure(self, client: TestClient):
        """Test that health response has correct structure."""
        response = client.get("/health")
        data = response.json()
        assert "status" in data
        assert "active_connections" in data
        assert data["status"] == "healthy"

    def test_health_active_connections_count(self, client: TestClient):
        """Test that active_connections is a non-negative integer."""
        response = client.get("/health")
        data = response.json()
        assert isinstance(data["active_connections"], int)
        assert data["active_connections"] >= 0

File: Rust/tests/test_runs_endpoint.rs
mod common;

use axum::http::StatusCode;
use axum_test::TestServer;
use common::{build_test_app, valid_payload_with};

#[tokio::test]
async fn test_empty_runs() {
    let server = TestServer::new(build_test_app()).unwrap();
    let response = server.get("/api/v1/runs").await;
    assert_eq!(response.status_code(), StatusCode::OK);
    let body: serde_json::Value = response.json();
    assert!(body.as_object().unwrap().is_empty());
}

#[tokio::test]
async fn test_single_run() {
    let server = TestServer::new(build_test_app()).unwrap();

    server
        .post("/api/v1/metrics/layerwise")
        .json(&valid_payload_with("run1", 100))
        .await;

    let response = server.get("/api/v1/runs").await;
    assert_eq!(response.status_code(), StatusCode::OK);
    let body: serde_json::Value = response.json();
    assert!(body.as_object().unwrap().contains_key("run1"));

    let run_info = &body["run1"];
    assert!(run_info["created_at"].is_string());
    assert!(run_info["last_update"].is_string());
    assert_eq!(run_info["step_count"], 1);
    assert_eq!(run_info["latest_step"], 100);
}

#[tokio::test]
async fn test_multiple_runs() {
    let server = TestServer::new(build_test_app()).unwrap();

    server
        .post("/api/v1/metrics/layerwise")
        .json(&valid_payload_with("run1", 100))
        .await;
    server
        .post("/api/v1/metrics/layerwise")
        .json(&valid_payload_with("run2", 200))
        .await;
    server
        .post("/api/v1/metrics/layerwise")
        .json(&valid_payload_with("run3", 300))
        .await;

    let response = server.get("/api/v1/runs").await;
    let body: serde_json::Value = response.json();
    assert_eq!(body.as_object().unwrap().len(), 3);
    assert!(body.as_object().unwrap().contains_key("run1"));
    assert!(body.as_object().unwrap().contains_key("run2"));
    assert!(body.as_object().unwrap().contains_key("run3"));
}

#[tokio::test]
async fn test_run_info_fields() {
    let server = TestServer::new(build_test_app()).unwrap();

    server
        .post("/api/v1/metrics/layerwise")
        .json(&valid_payload_with("run1", 100))
        .await;

    let response = server.get("/api/v1/runs").await;
    let body: serde_json::Value = response.json();
    let run_info = &body["run1"];

    assert!(run_info.get("created_at").is_some());
    assert!(run_info.get("last_update").is_some());
    assert!(run_info.get("step_count").is_some());
    assert!(run_info.get("latest_step").is_some());

    assert!(run_info["created_at"].is_string());
    assert!(run_info["last_update"].is_string());
    assert!(run_info["step_count"].is_number());
    assert!(run_info["latest_step"].is_number());
}

#[tokio::test]
async fn test_get_run_by_id() {
    let server = TestServer::new(build_test_app()).unwrap();

    server
        .post("/api/v1/metrics/layerwise")
        .json(&valid_payload_with("test_run", 100))
        .await;

    let response = server.get("/api/v1/runs/test_run").await;
    assert_eq!(response.status_code(), StatusCode::OK);
    let body: serde_json::Value = response.json();

    assert!(body["created_at"].is_string());
    assert!(body["last_update"].is_string());
    assert!(body["steps"].is_array());
}

#[tokio::test]
async fn test_get_run_not_found() {
    let server = TestServer::new(build_test_app()).unwrap();
    let response = server.get("/api/v1/runs/nonexistent").await;
    assert_eq!(response.status_code(), StatusCode::NOT_FOUND);
    let body: serde_json::Value = response.json();
    assert!(body["detail"]["error"].is_string());
    assert!(body["detail"]["message"].is_string());
}

#[tokio::test]
async fn test_latest_step() {
    let server = TestServer::new(build_test_app()).unwrap();

    server
        .post("/api/v1/metrics/layerwise")
        .json(&valid_payload_with("r1", 100))
        .await;
    server
        .post("/api/v1/metrics/layerwise")
        .json(&valid_payload_with("r1", 200))
        .await;

    let response = server.get("/api/v1/runs/r1/latest").await;
    assert_eq!(response.status_code(), StatusCode::OK);
    let body: serde_json::Value = response.json();
    assert_eq!(body["step"], 200);
}

#[tokio::test]
async fn test_latest_step_not_found() {
    let server = TestServer::new(build_test_app()).unwrap();
    let response = server.get("/api/v1/runs/nonexistent/latest").await;
    assert_eq!(response.status_code(), StatusCode::NOT_FOUND);
}

#[tokio::test]
async fn test_step_count_increments() {
    let server = TestServer::new(build_test_app()).unwrap();

    server
        .post("/api/v1/metrics/layerwise")
        .json(&valid_payload_with("r1", 100))
        .await;
    server
        .post("/api/v1/metrics/layerwise")
        .json(&valid_payload_with("r1", 200))
        .await;
    server
        .post("/api/v1/metrics/layerwise")
        .json(&valid_payload_with("r1", 300))
        .await;

    let response = server.get("/api/v1/runs").await;
    let body: serde_json::Value = response.json();
    assert_eq!(body["r1"]["step_count"], 3);
}

#[tokio::test]
async fn test_step_count_with_duplicate() {
    let server = TestServer::new(build_test_app()).unwrap();

    server
        .post("/api/v1/metrics/layerwise")
        .json(&valid_payload_with("r1", 100))
        .await;
    server
        .post("/api/v1/metrics/layerwise")
        .json(&valid_payload_with("r1", 100))
        .await;

    let response = server.get("/api/v1/runs").await;
    let body: serde_json::Value = response.json();
    assert_eq!(body["r1"]["step_count"], 1);
}

#[tokio::test]
async fn test_runs_after_multiple_posts() {
    let server = TestServer::new(build_test_app()).unwrap();

    server
        .post("/api/v1/metrics/layerwise")
        .json(&valid_payload_with("r1", 300))
        .await;
    server
        .post("/api/v1/metrics/layerwise")
        .json(&valid_payload_with("r1", 100))
        .await;
    server
        .post("/api/v1/metrics/layerwise")
        .json(&valid_payload_with("r1", 200))
        .await;

    let response = server.get("/api/v1/runs/r1").await;
    assert_eq!(response.status_code(), StatusCode::OK);
    let body: serde_json::Value = response.json();
    let steps = body["steps"].as_array().unwrap();
    assert_eq!(steps.len(), 3);
    assert_eq!(steps[0]["step"], 100);
    assert_eq!(steps[1]["step"], 200);
    assert_eq!(steps[2]["step"], 300);
}

#[tokio::test]
async fn test_health_endpoint() {
    let server = TestServer::new(build_test_app()).unwrap();
    let response = server.get("/health").await;
    assert_eq!(response.status_code(), StatusCode::OK);
    let body: serde_json::Value = response.json();
    assert_eq!(body["status"], "healthy");
    assert_eq!(body["active_connections"], 0);
}

#[tokio::test]
async fn test_health_active_connections_type() {
    let server = TestServer::new(build_test_app()).unwrap();
    let response = server.get("/health").await;
    let body: serde_json::Value = response.json();
    assert!(
        body["active_connections"].is_number(),
        "active_connections should be a number"
    );
    assert_eq!(body["active_connections"].as_u64().unwrap(), 0);
}

#[tokio::test]
async fn test_run_data_has_steps() {
    let server = TestServer::new(build_test_app()).unwrap();

    server
        .post("/api/v1/metrics/layerwise")
        .json(&valid_payload_with("r1", 100))
        .await;

    let response = server.get("/api/v1/runs/r1").await;
    let body: serde_json::Value = response.json();
    let steps = body["steps"].as_array().expect("steps should be array");
    assert_eq!(steps.len(), 1);
    assert_eq!(steps[0]["step"], 100);
    assert!(steps[0]["timestamp"].is_number());
    assert!(steps[0]["batch_size"].is_number());
}

#[tokio::test]
async fn test_cors_headers_present_on_get_runs() {
    let server = TestServer::new(build_test_app()).unwrap();

    let response = server
        .get("/api/v1/runs")
        .add_header("Origin", "https://example.com")
        .await;

    assert_eq!(response.status_code(), StatusCode::OK);
    assert!(
        response
            .headers()
            .contains_key("access-control-allow-origin"),
        "CORS Access-Control-Allow-Origin header should be present"
    );
    assert_eq!(
        response.headers()["access-control-allow-origin"]
            .to_str()
            .unwrap(),
        "*",
        "CORS origin should allow all origins (test app is permissive)"
    );
}

#[tokio::test]
async fn test_latest_step_has_layers() {
    let server = TestServer::new(build_test_app()).unwrap();

    server
        .post("/api/v1/metrics/layerwise")
        .json(&valid_payload_with("r1", 100))
        .await;

    let response = server.get("/api/v1/runs/r1/latest").await;
    assert_eq!(response.status_code(), StatusCode::OK);
    let body: serde_json::Value = response.json();
    assert!(body["layers"].is_array());
    assert!(!body["layers"].as_array().unwrap().is_empty());
}
